packages/appsync-modelgen-plugin/src/utils/process-belongs-to.ts

```diff
@@ -13,58 +14,67 @@ export function processBelongsToConnection(
   model: CodeGenModel,
   modelMap: CodeGenModelMap,
   connectionDirective: CodeGenDirective,
+  isCustomPKEnabled: boolean = false,
 ): CodeGenFieldConnection | undefined {
-  const otherSide = modelMap[field.type];
-  const otherSideField = getConnectedFieldV2(field, model, otherSide, connectionDirective.name);
-  const connectionFields = connectionDirective.arguments.fields || [];
-
-  if (connectionFields.length > 1) {
-    // Todo: Move to a common function and update the error message
-    throw new Error('DataStore only support one key in field');
-  }
-
   if (field.isList) {
     throw new Error(
       `A list field does not support the 'belongsTo' relation`
     );
   }
-
-  let validOtherSideField = false;
-  otherSideField.directives.forEach(dir => {
-    if (dir.name === 'hasOne' || dir.name === 'hasMany') {
-      validOtherSideField = true;
-    }
-  });
-
-  if (!validOtherSideField) {
+  const otherSide = modelMap[field.type];
+  const otherSideConnectedFields = getBelongsToConnectedFields(model, otherSide);
+  if (otherSideConnectedFields.length === 0) {
     throw new Error(
       `A 'belongsTo' field should match to a corresponding 'hasMany' or 'hasOne' field`
     );
   }
+  const otherSideField = isCustomPKEnabled ? otherSideConnectedFields[0] : getConnectedFieldV2(field, model, otherSide, connectionDirective.name);
+  const connectionFields = connectionDirective.arguments.fields || [];
+
   // if a type is connected using name, then amplify-graphql-relational-transformer adds a field to
   //  track the connection and that field is not part of the selection set
   // but if the field are connected using fields argument in connection directive
   // we are reusing the field and it should be preserved in selection set
   const otherSideHasMany = otherSideField.isList;
   const isConnectingFieldAutoCreated = false;
 
+  // New metada type introduced by custom PK v2 support
+  let targetNames: string[] = [ ...connectionFields ];
+  if (targetNames.length === 0) {
+    if (otherSideHasMany) {
+      targetNames = isCustomPKEnabled
+        ? getModelPrimaryKeyComponentFields(otherSide).map(componentField => makeConnectionAttributeName(otherSide.name, otherSideField.name, componentField.name))
+        : [makeConnectionAttributeName(otherSide.name, otherSideField.name)];
+    }
+    else {
+      targetNames = isCustomPKEnabled
+        ? getModelPrimaryKeyComponentFields(otherSide).map(componentField => makeConnectionAttributeName(model.name, field.name, componentField.name))
+        : [makeConnectionAttributeName(model.name, field.name)];
+    }
+  }
+
   return {
     kind: CodeGenConnectionType.BELONGS_TO,
     connectedModel: otherSide,
     isConnectingFieldAutoCreated,
-    targetName: connectionFields[0] || (otherSideHasMany ? makeConnectionAttributeName(otherSide.name, otherSideField.name) :
-      makeConnectionAttributeName(model.name, field.name)),
+    targetName: targetNames[0],
```

Comment:
since we are using the first item in `targetNames`, can we just get rid of the `targetName`?

Reply:
The `targetName` will be deprecated once all platforms consume this change